Dag init from counts

MBrouns opened the pull request:
Currently, dag construction works by supplying a dataframe where rows are duplicated `n` times where `n / len(df)` is the probability of occuring:
```
study | hard | grade
  -- |  -- | --
lots | yes | good
lots | yes | good
lots | yes | good
lots | yes | good
lots | yes | good
lots | yes | good
```

Another method could be to initialize by counts:
```
study | hard | grade | count
-- | -- | -- | --
little | no | fail | 20
little | no | good | 60
little | no | pass | 20
```
or by probabilities
```
study | hard | grade | prob
-- | -- | -- | --
little | no | fail | 0.033333
little | no | good | 0.100000
little | no | pass | 0.033333
little | yes | fail | 0.066667
little | yes | good | 0.050000
```
Initializing with probabilities would have two main downsides:
- floating points
- we lose information (no dirichlet type fun things)

so using counts seems to make more sense. Advantages of this approach are:
- smaller dataframe in the dag object
- should remove the need for the `cache` call in the counterfactuals api as we can modify the counts directly

Approach: 
- [x] reduce coupling with df attribute of dag object
- [ ] replace current `__init__` to accept  counts
- [ ] add `from_observations()` classmethod
- [ ] change counterfactuals inference to use the counts df

MBrouns committed:
reduce tight coupling with df implementation of dag object
- rename `self.df` to `self._df`
- add `values_for_node` method

MBrouns committed:
ensure `Dag().copy()` returns a copy of the graph object rather than a reference

MBrouns committed:
fix issue in `inference_dag` where edges to nodes that had a `do` operator on them would still show up in the graph used for inference

MBrouns committed:
fix mistake in _check_query_input

koaning commented on brent/graph.py:
````diff
@@ -45,9 +45,9 @@ def __init__(self, dataframe: pd.DataFrame):
         ```
         """
-        self.df = dataframe
+        self._df = dataframe
````
yep. good one.

koaning replied:
come to think of it, why don't we do a count here? i cannot come up with a single good reason why we'd need every single row here.

MBrouns replied:
I also think we don't ever need every single row, but wasn't sure how `df` was being used throughout the other parts of the code. That's why I decoupled a bit first. 

does `Dag.from_observations(df)` sound like a good method from where to do the `group_by().count()`?

koaning replied:
sure.

koaning commented on brent/graph.py:
```diff
@@ -86,7 +86,11 @@ def marginal_table(self):
     def nodes(self):
         """The nodes of the graph."""
-        return list(self.graph.nodes)
+        return set(self.graph.nodes)
```
also: good one.

koaning commented on brent/graph.py:
````diff
@@ -269,6 +274,33 @@ def add_edge(self, source, sink) -> 'DAG':
+            .remove_edge("a", "b"))
+        ```
+        """
+        try:
````
i might prefer a check (and a test). if the DAG is cached then this action should no longer be allowed.

koaning replied:
i'm not super sure what is great behavior with regards to that cacheing tho. might need to discuss this in person.

MBrouns replied:
Good point, I'll add a test. Let's discuss caching in person indeed

koaning commented:
i've merged v2 of counterfactuals yesterday by the way.

MBrouns committed:
Change main dataframe in DAG object to be an aggregated dataframe containing counts

- changes constructor to accept a dataframe with a count column
- adds `from_observations` classmethod for the old style initialization

koaning commented:
so the counterfactual api, turns out, is fundamentally broken. feel free to ignore anything from that api.
